File: interface/client/controllers/MarketController.ts
// Copyright © Spatial Corporation. All rights reserved.

import { Coin } from "@sptlco/data";
import { Controller } from "..";

/**
 * A controller for market functions.
 */
export class MarketController extends Controller {
  /**
   * Fetch the current market data.
   * @returns The current market data.
   */
  public current = async () => {
    return this.get<Coin>("market");
  };

  /**
   * Get the current account balance.
   * @returns The current account balance.
   */
  public balance = async () => {
    return this.get<string>("market/balance");
  };
}

File: cloud/Server/API/MarketController.cs
// Copyright © Spatial Corporation. All rights reserved.

using Nethereum.Web3;
using Spatial.Blockchain;
using Spatial.Blockchain.Helpers;
using Spatial.Cloud.Data.Scopes;
using Spatial.Identity.Authorization;

namespace Spatial.Cloud.API;

/// <summary>
/// A <see cref="Controller"/> for market functions.
/// </summary>
[Path("market")]
public class MarketController : Controller
{
    /// <summary>
    /// Fetch the current market data.
    /// </summary>
    /// <returns>The current market data.</returns>
    [GET]
    [Authorize(Scope.Market.Fetch)]
    public async Task<CoinGecko.Coin> FetchAsync()
    {
        return (await CoinGecko.GetMarketsAsync(["ethereum"]))[0];
    }

    /// <summary>
    /// Get the current Ethereum balance.
    /// </summary>
    /// <returns>The current Ethereum balance.</returns>
    [GET]
    [Path("balance")]
    [Authorize(Scope.Market.Fetch)]
    public async Task<string> GetBalanceAsync()
    {
        return Web3.Convert.FromWei(await Ethereum.CreateClient().GetBalanceAsync()).ToString();
    }
}